### ISSUE TITLE
Failed to build glib due to iconv error

### ISSUE THREAD

**Issue by weijianwen** [spack/spack#6962]
I failed to build glib on CentOS 7 due to error in configure: "GNU libiconv not in use but included iconv.h is from libiconv". GCC 4.8.5 shipped with CentOS is used.

### Expected Result

Spack installs glib.

### Actual Result

Spack failed to install glib due to iconv error.

### Steps to reproduce the issue

```console
$ spack install glib@2.55.1 %gcc@4.8.5
...
==> Error: ProcessError: Command exited with status 2:
    'make' '-j4'

6 errors found in build log:
     [ ... ]
     597     CC       libglib_2_0_la-gcharset.lo
     598     CC       libglib_2_0_la-gchecksum.lo
     599     CC       libglib_2_0_la-gconvert.lo
     600     CC       libglib_2_0_la-gdataset.lo
     601     CC       libglib_2_0_la-gdate.lo
     602     CC       libglib_2_0_la-gdatetime.lo
  >> 603   gconvert.c:59:2: error: #error GNU libiconv not in use but included iconv.h is from libiconv
     604    #error GNU libiconv not in use but included iconv.h is from libiconv
     605     ^
  >> 606   make[4]: *** [libglib_2_0_la-gconvert.lo] Error 1
     607   make[4]: *** Waiting for unfinished jobs....
     608   make[4]: Leaving directory `/lustre/home/acct-medhjy/public/spack/var/spack/stage/glib-2.55.1-wwkm635eqb5rmou57pu3n6ype2fcrj3e/glib-2.55.1/glib'
  >> 609   make[3]: *** [all-recursive] Error 1
     610   make[3]: Leaving directory `/lustre/home/acct-medhjy/public/spack/var/spack/stage/glib-2.55.1-wwkm635eqb5rmou57pu3n6ype2fcrj3e/glib-2.55.1/glib'
  >> 611   make[2]: *** [all] Error 2
     612   make[2]: Leaving directory `/lustre/home/acct-medhjy/public/spack/var/spack/stage/glib-2.55.1-wwkm635eqb5rmou57pu3n6ype2fcrj3e/glib-2.55.1/glib'
  >> 613   make[1]: *** [all-recursive] Error 1
     614   make[1]: Leaving directory `/lustre/home/acct-medhjy/public/spack/var/spack/stage/glib-2.55.1-wwkm635eqb5rmou57pu3n6ype2fcrj3e/glib-2.55.1'
  >> 615   make: *** [all] Error 2
```

### Information on your system

GCC 4.8.5 shipped with CentOS 7.3 is used. No relevant configuration on glib is made. Detailed logs are attached.

[spack-build.env.txt](https://github.com/spack/spack/files/1637542/spack-build.env.txt)
[spack-build.out.txt](https://github.com/spack/spack/files/1637543/spack-build.out.txt)

**Comment by weijianwen** [comment]
Adding `libiconv` and enabling it in `glib` configuring fixes this issue. I'll send a PR later.

**Comment by nazavode** [comment]
I tried this test inside a clean `centos7` system (via [this Dockerfile](https://gist.github.com/nazavode/87d073832894e71a7cf12db06e6b6aed)) and I cannot reproduce the issue, the spec `glib@2.55.1 %gcc@4.8.5` with `develop` branch at `4dd6ee25d609f87864cef71abfcb8ebe25516131` builds and installs just fine. Are you sure that `libiconv` dependency is mandatory for `glib`?

**Comment by weijianwen** [comment]
I reclone Spack and `glib` can be built successfully. Compared to the original Spack, the only differences are that `@system` openssl, ncurses and readline libraries are used. Specs between them are show as below.

The failing built glib
```console
$ spack spec glib %gcc@4.8.5
Input spec
--------------------------------
glib%gcc@4.8.5

Concretized
--------------------------------
glib@2.55.1%gcc@4.8.5~libmount patches=c325997b72a205ad1638bb3e3ba0e5b73e3d32ce63b2d0d3282f3e3a2ff4663c arch=linux-centos7-x86_64 
    ^gettext@0.19.8.1%gcc@4.8.5+bzip2+curses+git~libunistring+libxml2 patches=9acdb4e73f67c241b5ef32505c9ddf7cf6884ca8ea661692f21dca28483b04b8 +tar+xz arch=linux-centos7-x86_64 
        ^bzip2@1.0.6%gcc@4.8.5+shared arch=linux-centos7-x86_64 
        ^libxml2@2.9.4%gcc@4.8.5~python arch=linux-centos7-x86_64 
            ^pkgconf@1.4.0%gcc@4.8.5 arch=linux-centos7-x86_64 
            ^xz@5.2.3%gcc@4.8.5 arch=linux-centos7-x86_64 
            ^zlib@1.2.11%gcc@4.8.5+optimize+pic+shared arch=linux-centos7-x86_64 
        ^ncurses@system%gcc@4.8.5 patches=f84b2708a42777aadcc7f502a261afe10ca5646a51c1ef8b5e60d2070d926b57 ~symlinks~termlib arch=linux-centos7-x86_64 
        ^tar@1.29%gcc@4.8.5 patches=08921fcbd732050c74ddf1de7d8ad95ffdbc09f8b4342456fa2f6a0dd02a957c,5c314db58d005043bf407abaf25eb9823b9032a22fd12a0b142d4bf548130fa4,d428578be7fb99b831eb61e53b8d88a859afe08b479a21238180899707d79ce4,e34d2470461037933737c0a60fbcc5d5ea905fefb6b414b4a2e88a440fcae23a arch=linux-centos7-x86_64 
    ^libffi@3.2.1%gcc@4.8.5 arch=linux-centos7-x86_64 
    ^libiconv@1.15%gcc@4.8.5 arch=linux-centos7-x86_64 
    ^pcre@8.41%gcc@4.8.5~jit+utf arch=linux-centos7-x86_64 
    ^perl@5.24.1%gcc@4.8.5+cpanm+shared arch=linux-centos7-x86_64 
        ^gdbm@1.14.1%gcc@4.8.5 arch=linux-centos7-x86_64 
            ^readline@system%gcc@4.8.5 arch=linux-centos7-x86_64 
    ^python@2.7.14%gcc@4.8.5 patches=123082ab3483ded78e86d7c809e98a804b3465b4683c96bd79a2fd799f572244 +pic+shared~tk~ucs4 arch=linux-centos7-x86_64 
        ^openssl@system%gcc@4.8.5 arch=linux-centos7-x86_64 
        ^sqlite@3.21.0%gcc@4.8.5 patches=5bbcba091045e547eb550e6e9b9372009c41be0a9de8a5c030d03361c11ff939 arch=linux-centos7-x86_64
```

The successfully built glib
```console
$ spack spec glib %gcc@4.8.5
Input spec
--------------------------------
glib%gcc@4.8.5

Concretized
--------------------------------
glib@2.55.1%gcc@4.8.5~libmount patches=c325997b72a205ad1638bb3e3ba0e5b73e3d32ce63b2d0d3282f3e3a2ff4663c arch=linux-centos7-x86_64 
    ^gettext@0.19.8.1%gcc@4.8.5+bzip2+curses+git~libunistring+libxml2 patches=9acdb4e73f67c241b5ef32505c9ddf7cf6884ca8ea661692f21dca28483b04b8 +tar+xz arch=linux-centos7-x86_64 
        ^bzip2@1.0.6%gcc@4.8.5+shared arch=linux-centos7-x86_64 
        ^libxml2@2.9.4%gcc@4.8.5~python arch=linux-centos7-x86_64 
            ^pkgconf@1.4.0%gcc@4.8.5 arch=linux-centos7-x86_64 
            ^xz@5.2.3%gcc@4.8.5 arch=linux-centos7-x86_64 
            ^zlib@1.2.11%gcc@4.8.5+optimize+pic+shared arch=linux-centos7-x86_64 
        ^ncurses@6.0%gcc@4.8.5 patches=f84b2708a42777aadcc7f502a261afe10ca5646a51c1ef8b5e60d2070d926b57 ~symlinks~termlib arch=linux-centos7-x86_64 
        ^tar@1.29%gcc@4.8.5 patches=08921fcbd732050c74ddf1de7d8ad95ffdbc09f8b4342456fa2f6a0dd02a957c,5c314db58d005043bf407abaf25eb9823b9032a22fd12a0b142d4bf548130fa4,d428578be7fb99b831eb61e53b8d88a859afe08b479a21238180899707d79ce4,e34d2470461037933737c0a60fbcc5d5ea905fefb6b414b4a2e88a440fcae23a arch=linux-centos7-x86_64 
    ^libffi@3.2.1%gcc@4.8.5 arch=linux-centos7-x86_64 
    ^pcre@8.41%gcc@4.8.5~jit+utf arch=linux-centos7-x86_64 
    ^perl@5.24.1%gcc@4.8.5+cpanm+shared arch=linux-centos7-x86_64 
        ^gdbm@1.14.1%gcc@4.8.5 arch=linux-centos7-x86_64 
            ^readline@7.0%gcc@4.8.5 arch=linux-centos7-x86_64 
    ^python@2.7.14%gcc@4.8.5 patches=123082ab3483ded78e86d7c809e98a804b3465b4683c96bd79a2fd799f572244 +pic+shared~tk~ucs4 arch=linux-centos7-x86_64 
        ^openssl@1.0.2n%gcc@4.8.5 arch=linux-centos7-x86_64 
        ^sqlite@3.21.0%gcc@4.8.5 patches=5bbcba091045e547eb550e6e9b9372009c41be0a9de8a5c030d03361c11ff939 arch=linux-centos7-x86_64 
```

**Comment by nazavode** [comment]
I managed to reproduce your concretized spec by adding `libiconv` as an explicit dependency in `glib` (patch [here](https://gist.github.com/nazavode/b0ea66cd5b5ae2d80a58d8dcfe768281)) and running everything inside this [container](https://gist.github.com/nazavode/87d073832894e71a7cf12db06e6b6aed). In this case, the build fails as you reported:

```console
$ spack install glib@2.55.1 %gcc@4.8.5
[ ... ]
==> Error: ProcessError: Command exited with status 2:
    'make' '-j4'

6 errors found in build log:
     [ ... ]
  >> 595   gconvert.c:59:2: error: #error GNU libiconv not in use but included iconv.h is from libiconv
```

The error at `gconvert.c:59:2` is raised by the following preprocessor expression:

```c
#if !defined(USE_LIBICONV_GNU) && defined (_LIBICONV_H) \
     && !defined (__APPLE_CC__) && !defined (__LP_64__)
#error GNU libiconv not in use but included iconv.h is from libiconv
#endif
```

So it looks like that having `libiconv` available drives the `configure` to detect and enable the inclusion of `iconv.h` but according to this expression, to have that header *working properly* you have to explicitly enable `USE_LIBICONV_GNU`. I tried to reconfigure with  `--with-libiconv=gnu` and this time everything went fine.

Summing up:

1. `glib` build is ok as long as `libiconv` isn't available during the `configure` phase;
2. if you have `libiconv` in your spec, `glib` will fail to build unless `--with-libiconv=gnu` is specified during the `configure` phase.

I can help you with a PR if you want, but I'm asking the maintainers ( @alalazo @tgamblin ) here about best practices for situations like this one. IMO a default `~libiconv` variant on `glib` would do the trick, but I don't know if there is some better way to settle.

**Comment by weijianwen** [comment]
It seems to be my fault. Removing the dependency `depends_on('libiconv')` added by me then updated Spack to the latest version, the building goes well. It seems unneccesary to add libiconv dependency in glib.

Shall I close this issue?

**Comment by nazavode** [comment]
@weijianwen if you don't actually need `libiconv` then the current `spack` behaviour is ok, everything builds just fine, I think you can close the issue.
